analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/SymbolicCheckList.cs

```diff
@@ -60,19 +60,31 @@ public void ExecutionCompleted()
         }
 
         public SymbolicContext[] PreProcess(SymbolicContext context) =>
-            InvokeChecks(context, (check, context) => check.PreProcess(context));
+            InvokeChecks(context, preProcess: true);
 
         public SymbolicContext[] PostProcess(SymbolicContext context) =>
-            InvokeChecks(context, (check, context) => check.PostProcess(context));
+            InvokeChecks(context, preProcess: false);
 
-        private SymbolicContext[] InvokeChecks(SymbolicContext context, Func<SymbolicCheck, SymbolicContext, ProgramState[]> process)
+        private SymbolicContext[] InvokeChecks(SymbolicContext context, bool preProcess)
         {
-            var contexts = new[] { context };
+            var before = new List<SymbolicContext>() { context };
+            var after = new List<SymbolicContext>();
             foreach (var check in checks)
             {
-                contexts = contexts.SelectMany(x => process(check, x).Select(newState => x.WithState(newState))).ToArray();
+                foreach (var beforeContext in before)
+                {
+                    var newStates = preProcess ? check.PreProcess(beforeContext) : check.PostProcess(beforeContext);
+                    foreach (var newState in newStates)
+                    {
+                        after.Add(beforeContext.WithState(newState));
+                    }
+                }
+                var temp = before;
+                before = after;
+                after = temp;
```

Comment:
To consider: What about using `Interlocked.Exchange` to make this swap a single line? 
We don't really need the atomic aspect. It's just to consider